kw6/reader.py

```diff
@@ -143,68 +145,80 @@ def __getitem__(self, indices_or_slice):
                 ]
 
         elif isinstance(indices_or_slice, Iterable):
-            positions = [self.position_(index) for index in indices_or_slice]
+            positions = [self.seek_position_(index) for index in indices_or_slice]
 
         else:
             raise TypeError(f"Unindexable type {type(indices_or_slice)}")
 
         return positions
 
-    def position_(self, frame_index: types.FRAME_INDEX):
+    def seek_position_(self, frame_index: types.FRAME_INDEX):
         if frame_index < 0:
-            raise IndexError("Negative indexing not supported.")
-        try:
-            if self.inconsistent_position_bytes:
-                raise IOError
-            self.assumptuous_seek_(frame_index)
-            position = Position.from_stream_(self.stream)
-        except Exception:
-            self.inconsistent_position_bytes = True
-            self.seek_(frame_index)
-            if not self.empty():
-                position = Position.from_stream_(self.stream)
-            else:
-                raise IndexError(frame_index)
+            raise IndexError("Negative indexing not supported")
 
-        if position.header.frame_index != frame_index:
+        if frame_index < self.initial_frame_index:
             raise IndexError(
-                f"Expected to find {frame_index}, "
-                f"but found {position.header.frame_index} instead."
+                f"Frame index {frame_index} is smaller than the first frame index"
             )
 
-        return position
+        attempts = 0
+        position = None
+        step_size_confidence = -1
+        while not self.empty():
+            self.stream.seek(self.closest_stored_byte_position(frame_index))
+            current_frame_index = PositionHeader.peek_from_stream(self.stream).frame_index
+            try:
+                if step_size_confidence == -1:
+                    position = self.assumptuous_seek_position_(frame_index)
+                else:
+                    position = self.assumptuous_seek_position_(
+                        min(current_frame_index + step_size_confidence, frame_index)
+                    )
+                    step_size_confidence *= 10
+            except Exception:
```

Comment:
Might not want to blanket except `Exception` since that also captures things like `KeyboardInterrupt`

Reply:
True but we don't really know all the types of errors that we could get :)

Reply:
Okay then I guess we can leave it